contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -1639,3 +1666,275 @@ pltsql_replace_non_determinstic(text *src_text, text *from_text, text *to_text,
 	}
 	return false;
 }
+
+/*
+ * Find the matched length for substr in src_text
+ * Only matches if substr is prefix of src_text
+ */
+static bool
+icu_find_matched_length(char *src_text, int src_len, char *substr_text, int substr_len, Oid collid, int *matched_len, bool is_cs_ai)
+{
+    pg_locale_t mylocale = 0;
+
+    if (!lc_collate_is_c(collid))
+        mylocale = pg_newlocale_from_collation(collid);
+
+    if (!pg_locale_deterministic(mylocale) && mylocale->provider == 'i')
+    {
+#ifdef USE_ICU
+		int32_t src_len_utf8 = src_len;
+		int32_t substr_len_utf8 = substr_len;
+		int32_t src_ulen, substr_ulen;
+		int32_t u16_pos, u8_pos = 0;
+		UErrorCode  status = U_ZERO_ERROR;
+		UStringSearch *usearch;
+		UChar *src_uchar, *substr_uchar;
+		int32 matched_length_u16, u8_endpos;
+
+		src_ulen = icu_to_uchar(&src_uchar, src_text, src_len_utf8);
+		substr_ulen = icu_to_uchar(&substr_uchar, substr_text, substr_len_utf8);
+
+		usearch = usearch_openFromCollator(substr_uchar,
+										substr_ulen,
+										src_uchar,
+										src_ulen,
+										mylocale->info.icu.ucol,
+										NULL,
+										&status);
+		if (U_FAILURE(status))
+			ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+						errmsg("failed to perform ICU search: %s",
+							   u_errorName(status))));
+
+		usearch_setAttribute(usearch, USEARCH_OVERLAP, USEARCH_ON, &status);
+
+		/* substr should start matching from the first position in src string */
+		u16_pos = usearch_preceding(usearch, 1, &status);
+		if (!U_FAILURE(status) && u16_pos != USEARCH_DONE && (!is_cs_ai ||
+		    icu_compare_utf8_coll(mylocale->info.icu.ucol, src_uchar, usearch_getMatchedLength(usearch), substr_uchar, substr_len_utf8, false) == 0))
+		{
+			u8_pos = translate_char_pos(src_text, src_len_utf8,
```

Comment:
if u16_pos will always be zero at this point then we can probably skip this

Reply:
fixed